File: 2015/Day_08/Python/part_2.py
import re
import sys
sys.path.insert(0, '../../input_parsing')
import parse_input


def string_code_length(line: str) -> int:
    """Take in a string and return the number of raw characters in the string."""
    return len(line)


def encoded_strings(line: str) -> int:
    line_length = len(line)
    encode_count = 2
    for character in line:
        if character == '"' or character == "\\":
            encode_count += 1
    return line_length + encode_count


if __name__ == "__main__":
    santa_list = parse_input.input_per_line('../input.txt')
    answer = sum([encoded_strings(line) - string_code_length(line) for line in santa_list])
    print(f"The answer is: {answer}")

# 6802 is to high

File: 2015/Day_08/Ruby/part_2.rb
require "../../input_parsing/parse_input"

def count_code_rep(line)
    line.length
end

def count_encode(line)
    line_length = line.length
    encode_count = 2  # this is for the extra double-quotes added as part of the part2 encoding scheme
    chars_that_need_backslash = line.count '"\\'
    line_length + encode_count + chars_that_need_backslash
end

if $PROGRAM_NAME == __FILE__
    strings = input_per_line('../input.txt')
    code_rep_length = []
    encode_string_length = []
    strings.each do | line |
        code_rep_length.append(count_code_rep(line))
        encode_string_length.append(count_encode(line))
    end
    sum_code_rep = code_rep_length.reduce(0) { |sum, num| sum + num}
    sum_encode_string = encode_string_length.reduce(0) { |sum, num| sum + num}
    puts "Answer is #{sum_encode_string - sum_code_rep}"
end
